plugins/window-rules/window-rules.cpp

```diff
@@ -0,0 +1,234 @@
+#include <algorithm>
+#include <cfloat>
+#include <memory>
+#include <vector>
+
+#include <wayfire/plugin.hpp>
+#include <wayfire/view.hpp>
+#include <wayfire/view-access-interface.hpp>
+#include <wayfire/signal-definitions.hpp>
+#include <wayfire/view-transform.hpp>
+#include <wayfire/parser/rule_parser.hpp>
+#include <wayfire/lexer/lexer.hpp>
+#include <wayfire/variant.hpp>
+#include <wayfire/rule/lambda_rule.hpp>
+#include <wayfire/rule/rule.hpp>
+#include <wayfire/util/log.hpp>
+
+#include "lambda-rules-registration.hpp"
+#include "view-action-interface.hpp"
+
+class wayfire_window_rules_t : public wf::plugin_interface_t
+{
+  public:
+    void init() override;
+    void fini() override;
+    void apply(const std::string & signal, wf::signal_data_t *data);
+
+  private:
+    wf::lexer_t _lexer;
+
+    wf::signal_callback_t _created;
+    wf::signal_callback_t _maximized;
+    wf::signal_callback_t _unmaximized;
+    wf::signal_callback_t _minimized;
+    wf::signal_callback_t _fullscreened;
+
+    std::vector<std::shared_ptr<wf::rule_t>> _rules;
+
+    wf::view_access_interface_t _access_interface;
+    wf::view_action_interface_t _action_interface;
+
+    nonstd::observer_ptr<wf::lambda_rules_registrations_t> _lambda_registrations;
+};
+
+void wayfire_window_rules_t::init()
+{
+    // Get the lambda rules registrations.
+    _lambda_registrations = wf::lambda_rules_registrations_t::get_instance();
+
+    // TODO: Remove debug/test code.
+    // Test rule
+    auto reg = std::make_shared<wf::lambda_rule_registration_t>();
+    reg->rule = "on created if title contains \"Alacritty\"";
+    reg->if_lambda = [] (std::string signal, wayfire_view view)
+    {
+        LOGW("test rule IF_LAMBDA! signal=", signal, ", view=", view->get_app_id());
+
+        return false;
+    };
+    reg->else_lambda = [] (std::string signal, wayfire_view view)
+    {
+        LOGW("test rule ELSE_LAMBDA! signal=", signal, ", view=",
+            view->get_app_id());
+
+        return false;
+    };
+    auto error = _lambda_registrations->register_lambda_rule("test_rule", reg);
+    if (error)
+    {
+        LOGE("Window-rules: Error while registering test rule.");
+    }
+
+    // Build rule list.
+    auto section = wf::get_core().config.get_section("window-rules");
+    for (auto opt : section->get_registered_options())
+    {
+        _lexer.reset(opt->get_value_str());
+        auto rule = wf::rule_parser_t().parse(_lexer);
+        if (rule != nullptr)
```

Comment:
The rule parser will spit out a detailed error, including a pointer to the symbol in the rule string that is wrong.

Unfortunately it will be spit out on std::cerr because the logging facilities are not available in wf-utils.

Reply:
Ok, at this stage it is important the user gets any error. Eventually, the logging should be moved to wf-utils.